### PR TITLE
feat: allow episodes selection in `run_parallel`

### PR DESCRIPTION
**Pull request by ramyamounir**
After running a full experiment with all episodes, it’s common to notice a specific episode or range of episodes in the W&B logs that show unusual or interesting performance. Until now, rerunning or debugging those specific episodes required rerunning the entire experiment and marking them for detailed logging.

This PR introduces the ability to select specific episodes (or ranges of episodes) to rerun directly. This makes targeted debugging and analysis faster and more reproducible, and it also enables running subsets of episodes for specialized tasks such as visualization or plotting.

I’ve validated that this approach reproduces the exact same performance as when the selected episodes were run as part of the full experiment.

### CLI Examples

```zsh
# Run all episodes (default behavior)
python run_parallel.py -e my_experiment

# Run a single episode
python run_parallel.py -e my_experiment -i "3"

# Run multiple specific episodes
python run_parallel.py -e my_experiment -i "0,2,5"

# Run a closed range (end exclusive)
python run_parallel.py -e my_experiment -i "3:7"

# Run an open-ended range (from start or to end)
python run_parallel.py -e my_experiment -i ":5"    # episodes 0–4
python run_parallel.py -e my_experiment -i "7:"    # episodes 7–end

# Combine multiple selections
python run_parallel.py -e my_experiment -i "0,3,5:8"
```

### CODE REVIEW

**Comment by ramyamounir** [comment]
FYI, @thousandbrainsproject/monty-maintainers

**Comment by tristanls-tbp** [comment]
**question**: Is there a reason this option is omitted from `run.py`?

**Comment by ramyamounir** [comment]
Unit tests added in [11ab20f8b891](https://github.com/ramyamounir/tbp.monty/commit/11ab20f8b8914357c0153233e9836acb5c82aae8)

> Is there a reason this option is omitted from `run.py`?

Yes, `run.py` is trickier to rerun specific episodes with reproducibility. The episodes run back to back without resetting the `rng`, so we cannot skip the earlier episodes because every `rng` call changes its internal state for the subsequent episodes. I think it's best to make sure the `rng` is always reset in `pre_episode` with the same seed, this way it would be easier to run specific episodes and expect the same results.

File: tests/unit/frameworks/run_parallel_test.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+
+class ParseEpisodeSpecTest(unittest.TestCase):
+    def test_selects_all(self):
+        self.assertEqual(parse_episode_spec(None, total=5), [0, 1, 2, 3, 4])
+        self.assertEqual(parse_episode_spec("", total=3), [0, 1, 2])
+        self.assertEqual(parse_episode_spec(":", total=3), [0, 1, 2])
+        self.assertEqual(parse_episode_spec("all", total=4), [0, 1, 2, 3])
+
+    # Single index tests
```
Comment (tristanls-tbp) [inline tests/unit/frameworks/run_parallel_test.py]: **issue**: extraneous comment, please remove

**note**: test names being `test_single_index_` obviate the need for the comment.

File: tests/unit/frameworks/run_parallel_test.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    def test_single_index_negative_raises(self):
+        with self.assertRaisesRegex(ValueError, "not a valid index"):
+            parse_episode_spec("-1", total=5)
+
+    def test_single_index_out_of_bounds_raises(self):
+        with self.assertRaisesRegex(ValueError, "not a valid index"):
+            parse_episode_spec("5", total=5)
+
+    # Closed range tests
```
Comment (tristanls-tbp) [inline tests/unit/frameworks/run_parallel_test.py]: **issue**: extraneous comment, please remove

**note**: test names being `test_closed_range_` obviate the need for the comment.

File: tests/unit/frameworks/run_parallel_test.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            parse_episode_spec("5:3", total=10)
+
+    def test_closed_range_end_equals_total_ok(self):
+        self.assertEqual(parse_episode_spec("0:4", total=4), [0, 1, 2, 3])
+
+    def test_closed_range_end_greater_than_total_raises(self):
+        with self.assertRaisesRegex(ValueError, "not a valid range"):
+            parse_episode_spec("0:6", total=5)
+
+    # Open range tests
```
Comment (tristanls-tbp) [inline tests/unit/frameworks/run_parallel_test.py]: **issue**: extreaneous comment, please remove

**note**: test names being `test_open_` obviate the need for the comment.

File: tests/unit/frameworks/run_parallel_test.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    def test_open_right_start_equals_total_invalid(self):
+        with self.assertRaisesRegex(ValueError, "not a valid range"):
+            parse_episode_spec("5:", total=5)
+
+    def test_open_right_start_negative_invalid(self):
+        with self.assertRaisesRegex(ValueError, "not a valid range"):
+            parse_episode_spec("-1:", total=5)
+
+    # Multiple selection and other edge cases
```
Comment (tristanls-tbp) [inline tests/unit/frameworks/run_parallel_test.py]: **issue**: extraneous comment, please remove

File: src/tbp/monty/frameworks/run_parallel.py [hunk 5]
```diff
@@ -469,6 +471,102 @@ def run_episodes_parallel(
...
+    if episode_spec is None:
+        return list(range(total))
+    s = episode_spec.strip().lower()
+    if s in ("", "all", ":"):
+        return list(range(total))
+
+    re_open_left = re.compile(r"^:([+-]?\d+)$")  # ":N"
+    re_open_right = re.compile(r"^([+-]?\d+):$")  # "N:"
+    re_closed = re.compile(r"^([+-]?\d+)\s*:\s*([+-]?\d+)$")  # "A:B"
+    re_single = re.compile(r"^[+-]?\d+$")  # "N"
```
Comment (tristanls-tbp) [inline src/tbp/monty/frameworks/run_parallel.py]: **question**: Why are `+` or `-` allowed? There is no use case for those.

**suggestion**: Remove

Comment (tristanls-tbp) [reply src/tbp/monty/frameworks/run_parallel.py]: **issue (non-blocking)**: I think these should be constants: `RE_OPEN_LEFT`, `RE_OPEN_RIGHT`, etc. and compiled outside of this function.

File: src/tbp/monty/frameworks/run_parallel.py [hunk 6]
```diff
@@ -646,6 +745,7 @@ def main(
             from command line as the config is selected from `all_configs`.
         experiment: Name of experiment to run. Not required if running
             from command line.
+        episodes: The episodes ids to run. Default is "all".
```
Comment (tristanls-tbp) [inline src/tbp/monty/frameworks/run_parallel.py]: **issue**: [Our convention is "Defaults to ..."](https://github.com/search?q=repo%3Athousandbrainsproject%2Ftbp.monty%20%22Defaults%20to%22&type=code).

File: src/tbp/monty/frameworks/run_parallel.py (hunk 5 above)
Comment (ramyamounir) [reply src/tbp/monty/frameworks/run_parallel.py]: I needed to catch the `+` and `-` and include them in the group so I can turn it in to `int` with the sign and apply the right checks/tests.

For example, if the user inputs `-i ":-5"` I want to include the `-` in the extracted group so I can apply the correct check and raise the correct error (incorrect range). If a user inputs `-i "+5:+8"`, I want the + to be included so that it can be later ignored when I cast it to `int`, otherwise it won't match any of the patterns defined.

File: tests/unit/frameworks/run_parallel_test.py (hunk 4 above)
Comment (ramyamounir) [reply tests/unit/frameworks/run_parallel_test.py]: Extraneous comments removed in [af17fd405101](https://github.com/ramyamounir/tbp.monty/commit/af17fd405101e8036c9f7c0e4821a53c88acb5e0)

File: src/tbp/monty/frameworks/run_parallel.py (hunk 5 above)
Comment (ramyamounir) [reply src/tbp/monty/frameworks/run_parallel.py]: Regex statements became constants outside the function in [342e4ab68008](https://github.com/ramyamounir/tbp.monty/commit/342e4ab680089dc2348175c3717d422980aa9ef4)

File: src/tbp/monty/frameworks/run_parallel.py (hunk 6 above)
Comment (ramyamounir) [reply src/tbp/monty/frameworks/run_parallel.py]: Fixed in [ad37a1df3ad3](https://github.com/ramyamounir/tbp.monty/commit/ad37a1df3ad31c3860f19a6c931fd6917d8054db)

**Comment by vkakerbeck** [comment]
Cool! This is a really useful option to have :) Also an interesting idea to just reset the rng in every pre-episode call. I hadn't thought of this before but it sounds like it would save us a lot of reproducibility headaches that we tried to work around with the follow-up configs script.

**Comment by ramyamounir** [comment]
@vkakerbeck , thanks! Yes, resetting `rng` at the beginning of each episode is now on my non-urgent Todo list. I'll get to it soon and then we could probably delete the follow-up configs script if we can reproduce specific episodes the same way with `run.py`.

File: src/tbp/monty/frameworks/run_parallel.py (hunk 5 above)
Comment (tristanls-tbp) [reply src/tbp/monty/frameworks/run_parallel.py]: > For example, if the user inputs -i ":-5" I want to include the - in the extracted group so I can apply the correct check and raise the correct error (incorrect range). If a user inputs -i "+5:+8", I want the + to be included so that it can be later ignored when I cast it to int, otherwise it won't match any of the patterns defined.

**issue**: Allowing both signed and unsigned formats increases complexity and potential for bugs without apparent user demand. None of the examples or tests so far use negative or signed values, such as `+5:+8`. Furthermore, `+5:+8` is not idiomatic, and I see no reason for its inclusion. Since users generally know the total episode count and all episodes are numbered with positive integers, requiring positive integers simplifies both usage and maintenance. Please remove support for `+` and `-`.

Comment (ramyamounir) [reply src/tbp/monty/frameworks/run_parallel.py]: In [34502a535590](https://github.com/ramyamounir/tbp.monty/commit/34502a5355904499548069cfd2b704b0d7faddee)
- Support for `+` and `-` removed. 
- Unit tests `test_single_index_negative_raises` and `test_open_right_start_negative_invalid` removed.

**Review by tristanls-tbp — APPROVED** [review]
Thank you.

File: src/tbp/monty/frameworks/run_parallel.py (hunk 6 above)
Comment (jeremyshoemaker) [reply src/tbp/monty/frameworks/run_parallel.py]: Putting the default in the docstring is redundant, similar to putting the type in the docstring. Does the generated documentation not include the defaults in the function signatures?

Comment (ramyamounir) [reply src/tbp/monty/frameworks/run_parallel.py]: Yeah, good point. I think, in another style PR, we can remove all `Defaults to ...` from the docstrings. The default value is easy to look up in the signature like you said.

That said, I think we're copying other documentation conventions where they specify the default value in both places (e.g., [numpy](https://numpy.org/devdocs/reference/generated/numpy.array.html#numpy-array))

Comment (tristanls-tbp) [reply src/tbp/monty/frameworks/run_parallel.py]: We are using the Google docstring convention. https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings

The `Defaults to` comes from Sphinx (what we use to generate our docs) example of Google docstrings: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
